src/java/org/apache/cassandra/io/sstable/format/bti/PartitionIndex.java

```diff
@@ -0,0 +1,452 @@
+ * </ul>
+ * To avoid having to create an object to carry the result, the two are distinguished by sign. Direct-to-dfile entries
+ * are recorded as ~position (~ instead of - to differentiate 0 in ifile from 0 in dfile).
+ * <p>
+ * In either case the contents of the file at this position start with a serialization of the key which can be used
+ * to verify the correct key is found.
+ * <p>
+ * The indexes are created by {@link PartitionIndexBuilder}. To read the index one must obtain a thread-unsafe
+ * {@link Reader} or {@link IndexPosIterator}.
+ */
+@VisibleForTesting
+public class PartitionIndex implements SharedCloseable
+{
+    private static final Logger logger = LoggerFactory.getLogger(PartitionIndex.class);
+
+    private final FileHandle fh;
+    private final long keyCount;
+    private final DecoratedKey first;
+    private final DecoratedKey last;
+    private final long root;
+
+    public static final long NOT_FOUND = Long.MIN_VALUE;
+    public static final int FOOTER_LENGTH = 3 * 8;
+
+    @VisibleForTesting
+    public PartitionIndex(FileHandle fh, long trieRoot, long keyCount, DecoratedKey first, DecoratedKey last)
+    {
+        this.keyCount = keyCount;
+        this.fh = fh.sharedCopy();
+        this.first = first;
+        this.last = last;
+        this.root = trieRoot;
+    }
+
+    private PartitionIndex(PartitionIndex src)
+    {
+        this(src.fh, src.root, src.keyCount, src.first, src.last);
+    }
+
+    static class Payload
+    {
+        final long position;
+        final short hashBits;
+
+        public Payload(long position, short hashBits)
+        {
+            this.position = position;
+            assert this.position != NOT_FOUND;
+            this.hashBits = hashBits;
+        }
+    }
+
+    static final PartitionIndexSerializer TRIE_SERIALIZER = new PartitionIndexSerializer();
+
+    private static class PartitionIndexSerializer implements TrieSerializer<Payload, DataOutputPlus>
+    {
+        public int sizeofNode(SerializationNode<Payload> node, long nodePosition)
+        {
+            return TrieNode.typeFor(node, nodePosition).sizeofNode(node) +
+                   (node.payload() != null ? 1 + SizedInts.nonZeroSize(node.payload().position) : 0);
+        }
+
+        @Override
+        public void write(DataOutputPlus dest, SerializationNode<Payload> node, long nodePosition) throws IOException
+        {
+            write(dest, TrieNode.typeFor(node, nodePosition), node, nodePosition);
+        }
+
+        public void write(DataOutputPlus dest, TrieNode type, SerializationNode<Payload> node, long nodePosition) throws IOException
+        {
+            Payload payload = node.payload();
+            if (payload != null)
+            {
+                int payloadBits;
+                int size = SizedInts.nonZeroSize(payload.position);
+                payloadBits = 7 + size;
```

Comment:
```suggestion
                int payloadBits = 7 + size;
```
nit: Just declare when it's assigned?

Reply:
Aside: So `size` is the number of bytes in the payload, but where does the `7` come from? (The hash is just one byte?)

Reply:
Added a comment and a `FLAG_HAS_HAS_BITS` constant.